include/cuda/memory_resource

```diff
@@ -0,0 +1,519 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the CUDA Toolkit, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_MEMORY_RESOURCE
+#define _CUDA_MEMORY_RESOURCE
+
+// clang-format off
+/*
+    memory_resource synopsis
+namespace cuda {
+namespace mr {
+template <class Resource>
+concept resource = equality_comparable<Resource>
+                && requires(Resource& res, void* ptr, size_t size, size_t alignment) {
+    res.allocate(size, alignment) -> void*;
+    res.deallocate(ptr, size, alignment) -> void;
```

Comment:
I don't know how important it is for this comment to be valid C++. Assuming it is:
```suggestion
    { res.allocate(size, alignment) } -> same_as<void*>;
    { res.deallocate(ptr, size, alignment) } -> same_as<void>;
```